arm-software/shared/elf2bin/elf2bin.h

```diff
@@ -0,0 +1,157 @@
+#include "llvm/Object/ELFObjectFile.h"
+#include "llvm/Support/Error.h"
+
+// A structure that describes a single loadable segment in an ELF
+// image. 'fileoffset' and 'filesize' designate a region of bytes in
+// the file; 'baseaddr' specifies the memory address that region is
+// expected to be loaded at.
+//
+// 'memsize' indicates how many bytes the region will occupy once it's
+// loaded. This should never be less than 'filesize'. If it's more
+// than 'filesize', it indicates that zero padding should be appended
+// by the loader to pad to the full size.
+struct Segment {
+  uint64_t baseaddr, fileoffset, filesize, memsize;
+};
+
+// A structure containing the name of an input file to the program,
+// and the result of loading it into an ELFObjectFile.
+struct InputObject {
+  std::string filename;
+  std::unique_ptr<llvm::MemoryBuffer> membuf;
+  std::unique_ptr<llvm::object::ELFObjectFileBase> elf;
+
+  // List the loadable segments in the file.
+  //
+  // The flag 'physical' indicates that the 'baseaddr' value for each
+  // returned segment should be the physical address of that segment,
+  // i.e. the p_paddr field in its program header table entry. If it's
+  // false, the virtual address will be used instead, i.e. the p_vaddr
+  // field.
+  std::vector<Segment> segments(bool physical);
+
+  uint64_t entry_point();
+};
+
+// Write a single binary or Verilog hex file. (A Verilog hex file is
+// just the plain binary data, represented as a sequence of text lines
+// each containing a hex-encoded byte).
+//
+// Input data comes from the file 'infile', at offset 'fileoffset',
+// 'size' bytes long. 'zi_size' zero bytes are appended after that.
+//
+// Bank switching is supported by the bank_* parameters, which select
+// a subset of the input bytes to be written to the output.
+// Specifically, each input byte is included or excluded depending on
+// the residue of its position in the input, mod 'bank_modulus'.
+// 'bank_nres' consecutive residues are kept, starting at
+// 'bank_firstres'. For example:
+//
+// modulus=8, firstres=0, nres=1: keep only the bytes whose positions
+// are congruent to 0 mod 8. (Just one residue, namely 0.) That is,
+// divide the input into 8-byte blocks and only write the initial byte
+// of each block.
+//
+// modulus=8, firstres=0, nres=2: keep the bytes whose positions are
+// congruent to {0,1} mod 8. (Two consecutive residues, starting at 0.)
+//
+// modulus=8, firstres=4, nres=2: keep the bytes whose positions are
+// congruent to {4,5} mod 8. (Still two residues, but now starting at 4.)
+//
+// If no bank switching is needed, then modulus=1, firstres=0, nres=1
+// is a combination that indicates 'write all input bytes to the output'.
+//
+// The output is written to the file 'outfile'.
+//
+// These functions will exit the entire program with an error message if
+// anything goes wrong. So callers need not handle the failure case.
+void bin_write(InputObject &inobj, const std::string &outfile,
```

Comment:
I couldn't spot any non-const uses of `inobj` on the implementation side of the `*_write` methods (I might have missed something, though). Can they be made a `const &` in these interfaces?